crates/ra_hir_ty/src/lib.rs

```diff
@@ -863,7 +863,7 @@ pub trait TypeWalk {
             &mut |ty, binders| {
                 if let &mut Ty::Bound(bound) = ty {
                     if bound.debruijn >= binders {
-                        *ty = substs.0[bound.index].clone();
+                        *ty = substs.0[bound.index].clone().shift_bound_vars(binders);
```

Comment:
Is there perhaps some way we could similar wrong code crash on all examples, and only just on the bad ones?

Reply:
Like, if this happens to work most of the time, b/c binders is `0`, can we remember a number which tells the nesting/how many times we should shift, and than assert everywhere that nesting levels match?

Reply:
I don't see how -- if I remember to assert, I can just do it correctly in that place. We could validate types to check that the binders are correct, but that will also only crash if we try to bind to indexes that don't exist.